cinema/views.py

```diff
@@ -48,6 +78,36 @@ class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
+    def get_queryset(self):
+        queryset = super().get_queryset()
+
+        date = self.request.query_params.get("date")
+
+        movie_id = self.request.query_params.get("movie")
+
+        if date:
+            queryset = queryset.filter(show_time__date=date)
+        if movie_id:
+            queryset = queryset.filter(movie__id=int(movie_id))
+
+        if self.action == "list":
+            queryset = (
+                queryset.select_related("movie", "cinema_hall")
+                .annotate(
+                    tickets_available=F(
+                        "cinema_hall__rows"
+                    ) * F(
+                        "cinema_hall__seats_in_row"
+                    ) - Count(
+                        "tickets"
+                    )
+                )
+            )
+        else:
+            queryset.select_related("movie", "cinema_hall")
```

Comment:
The `select_related` method is not being called correctly in the `else` block. It should be assigned back to `queryset` to ensure the query is optimized. Consider changing it to `queryset = queryset.select_related("movie", "cinema_hall")`.